### PR TITLE
Updated Criteo API version (2023-01 -> 2023-10)

### PR DESCRIPTION
**Pull request by MarcDillar**
## Testing

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

**Comment by joe-ayoub-segment** [comment]
Thanks for the PR @MarcDillar - I'll schedule for review.

**Comment by joe-ayoub-segment** [comment]
hi @MarcDillar thanks again for raising this PR. It looks good from my perspective. 

The main risk with a change like this is that we break data collection for customers who are already using the Integration. 

We don't have a capbility to deploy code dark, and it's not possible to do side by side testing either. So we're going to have to rely on some other mechanisms.  

I'm going to run an internal test to run some live traffic through it to see if any internal errors generated are the same as with the previous code. 

We could deploy the code to an internal Staging environment, and if you were free to jump on a call we could connect the Integration and sync an Audience to Criteo (which you could then verify looked good). Would you be open to that?

Best regards,
Joe

**Comment by joe-ayoub-segment** [comment]
note: meeting with Marc Dillar Monday

**Comment by joe-ayoub-segment** [comment]
Thanks for meeting earlier today @MarcDillar . 
Shall we schedule another call to test again later this week?
Here's my Calendly link: http://calendly.com/joe_ayoub/integration-workshop

**Comment by joe-ayoub-segment** [comment]
Hi @MarcDillar - looks like the CI checks are failing. 
image

**Comment by joe-ayoub-segment** [comment]
You can run the tests with this command: 

yarn jest --testPathPattern='./packages/destination-actions/src/destinations/criteo-audiences'

**Comment by MarcDillar** [comment]
Hi @joe-ayoub-segment 

Thank you, the unit tests should now work correctly:
image

**Comment by joe-ayoub-segment** [comment]
hi @MarcDillar  - in the end I decided to not deploy this PR until I am fully able to test it end to end locally without any potential issues. I raised a bug on our testing tool as I suspect there is an issue with it. Once this is done I'll retest and deploy. It will likely be in Jan 2024. 

Do we have enough time for this, or will the old API version stop working soon?

**Comment by MarcDillar** [comment]
Hi @joe-ayoub-segment, no worries, the endpoints that we had to replace are present until version 2023-07, which is supported until July 2024 so we still have a bit of time to make that change.

**Comment by joe-ayoub-segment** [comment]
Hi @MarcDillar , 

Happy New Year!
I'd like to get this PR deployed as soon as possible. I'm going to try doing a retest now - but just noticed that the linter is complaining about some 'var' definitions. I tried to fix myself but I don't have permission to push to your repo.  

image

**Comment by MarcDillar** [comment]
Hi @joe-ayoub-segment 

Happy new year to you too! I just pushed a commit that should solve the linting issue.

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]

File: packages/destination-actions/src/destinations/criteo-audiences/criteo-audiences.ts [hunk 1]
```diff
@@ -105,104 +105,124 @@ export const patchAudience = async (
...
+          advertiser_id
+        ]
+      }
+    }
+  }
 
-  const body = await response.json()
+  let continue_search = true
+  let offset = 0
+  do {
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/criteo-audiences/criteo-audiences.ts:134]: Finally discovered the bug which was causing requests to hang. The response sizes from the request to get the contactlist ID are too large. Approx 30kb. So Segment is hanging.

We can override this limit by adding the 'skipResponseCloning: true' attribute to the Request.

Also, we don't need to await response.json(). Instead we can do as below:

  interface AudienceSegment {
    attributes: {
      [key: string]: unknown
    }
    id: string
    type: string
  }

  interface ApiResponse {
    data: AudienceSegment[]
    meta: {
      totalItems: number
    }
  }

  let body: ApiResponse

  do {
    const endpoint = `${BASE_API_URL}/marketing-solutions/audience-segments/search?limit=${LIMIT}&offset=${offset}`

    const response = await request(endpoint, {
      method: 'POST',
      skipResponseCloning: true,
      headers: headers,
      json: payload
    })

    body = response.data as ApiResponse

    if (response.status !== 200)
      // Centrifuge will automatically retry the batch if there's
      // an issue fetching the Advertiser's audiences from Criteo.
      throw new RetryableError("Error while fetching the Advertiser's audiences")

    // If the contact list is found, return the corresponding ID
    for (const contactlist of body.data) {
      if (contactlist.attributes.name === audience_segment_name) return contactlist.id
    }

    // Else, continue searching
    offset += LIMIT
    continue_search = body.meta.totalItems > offset
  } while (continue_search)
}

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]

**Comment by joe-ayoub-segment** [comment]
hi @MarcDillar I did some more testing and found the issue which was causing the code to periodically hang.
I'm glad we didn't deploy it because it would have caused an incident :). 

Please see my comment above which should resolve the issue. Can you please update the PR accordingly?

While testing the PR I reviewed the code in more detail and noticed that there are a 3 potential failure points with the Integration code. 

1. We are fetching the contactlist ID with every invocation, which is inefficient. We do have a mechanism to cache these responses, but it will involve a more detailed change. 
2. The access_token is requested with each invocation, which is also inefficient. We do have a solution where we can cache the access_token as well, but that too will require a more in-depth change to the Integration code. 
3. Finally, I believe that the using a do..while loop to fetch the contactlist ID is inefficient. Each response can be large, and multiple requests may need to be made before the correct contactlist ID is retrieved. Would it not be possible to update the Criteo API to accept a contactlist Name and return a contactlist ID in a single request? I think we discussed this previously and you said that it's not possible yet. 

I look forward to hearing back from you. 
Kind regards,
Joe

**Comment by MarcDillar** [comment]
Hi @joe-ayoub-segment thank you for your review, I just made a commit following your recommendations.

**Review by joe-ayoub-segment — APPROVED** [review]

**Comment by joe-ayoub-segment** [comment]
Hi @MarcDillar this PR has been deployed. 

Could you please confirm that the change looks good from your side please? Is there any dashboard data you can check to verify this?

Best regards,
Joe

**Pull request merged**